casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -98,140 +67,122 @@ object Running {
+      casperAddF: BlockHash => F[Unit],
+      lockMap: TrieMap[BlockHash, Semaphore[F]]
+  ): F[Unit] =
+    for {
+      _ <- Log[F].info(
+            s"Incoming BlockMessage ${PrettyPrinter.buildString(blockMessage, short = true)} " +
+              s"from ${peer.endpoint.host}. Number of hashes currently handled is ${lockMap.size}."
           )
-          for {
-            _ <- Log[F].debug(
-                  s"Request ${PrettyPrinter.buildString(hash)} from ${nextPeer} " +
-                    s"remain waiting ${waitingListTail.mkString(",")}"
+      lock     <- MetricsSemaphore.single
+      hashLock <- lockMap.getOrElseUpdate(blockMessage.blockHash, lock).pure[F]
+      blockStr = PrettyPrinter.buildString(blockMessage, short = true)
+      proceed <- hashLock.withPermit(
+                  ignoreMessageF(blockMessage.blockHash)
+                    .flatMap(r => {
+                      if (r.doIgnore) {
+                        Log[F].info(
+                          s"Ignoring BlockMessage ${blockStr} because of ${r.status}."
+                        ) >> false.pure[F]
+                      } else {
+                        BlockStore[F]
+                          .contains(blockMessage.blockHash)
+                          .ifM(
+                            Log[F].info(
+                              s"Block ${blockStr} is available in BlockStore. Sending to Casper."
+                            ) >> true.pure[F],
+                            validateBlockF(blockMessage, peer).ifM(
+                              BlockStore[F].put(blockMessage) >>
+                                Log[F].info(
+                                  s"Block ${blockStr} is good. Saved to BlockStore. Sending to Casper."
+                                ) >> true.pure[F],
+                              Log[F].warn(
+                                s"Ignoring BlockMessage ${blockStr} because its malformed or not of our interest."
+                              ) >> false.pure[F]
+                            )
+                          )
+                      }
+                    })
+                    .ifM(
+                      BlocksRetriever[F].ackReceive(blockMessage.blockHash) >> true.pure[F],
+                      false.pure[F]
+                    )
                 )
-            _  <- requestForBlock(nextPeer, hash)
-            ts <- Time[F].currentMillis
-          } yield hash -> Option(modifiedRequested(ts))
-        case _ =>
-          val warnMessage = s"Could not retrieve requested block ${PrettyPrinter.buildString(hash)} " +
-            s"from ${requested.peers.mkString(",")}. Removing the request from the requested blocks list. " +
-            s"Casper will have to re-request the block."
-          Metrics[F].incrementCounter("block-retrieve-failed") >>
-            Log[F].warn(warnMessage).as(hash -> none[Requested])
-      }
+      _ <- lockMap.remove(blockMessage.blockHash).pure[F]
```

Comment:
In a situation where multiple threads enter this method. The first will start execution and others will wait. After the first is finished, the lock will be removed. The same lock for the one currently executing (entered second) and others that are still waiting. So now `lockMap` does not have lock for this block and next request will start execution immediately in parallel.

Reply:
This is the intention here - we just need to hold all handlers for particular block except for one that should go through the code inside lock and verify it. Then, when the hoard will go forward because there is in lock - they should be stopped on `ignoreMessageF` call

Reply:
Thanks for highlighting this, actually when blocks are all invalid, we have to validate them all. But idk whats the best way to mitigate this attack vector.

Reply:
Oh, I see. Casper is not involved here, just block validation and saving to store.